### PR TITLE
[Stats] Metrics Export User Interface Part 2 (Prometheus Service Discovery)

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

The dashboard will run a thread that keeps writing a prometheus service discovery file. 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/metrics_agent.py [hunk 1]
```diff
@@ -181,3 +187,65 @@ def _register_if_needed(self, metric_point: MetricPoint):
...
+                metrics_export_address
+                for metrics_export_address in metrics_export_addresses
+            ]
+        }]
+        with open(self.get_target_file_name(), "w") as json_file:
+            json.dump(data, json_file)
+
+    def get_target_file_name(self):
+        return os.path.join(
+            self.temp_dir, ray.ray_constants.PROMETHEUS_SERVICE_DISCOVERY_FILE)
```
Comment (simon-mo) [inline python/ray/metrics_agent.py:217]: This doesn't need to be in its own class. The abstraction and it's interaction with exception handling in the while True block in the thread is leaky.

**Review by simon-mo — COMMENTED** [review]
Nice!

File: python/ray/metrics_agent.py (hunk 1 above)
Comment (rkooo567) [reply python/ray/metrics_agent.py:217]: Right. That's a good point. Let me merge two classes.

**Review by edoakes — APPROVED** [review]
Looks good. My only concern is about the test - ideally we would mock out the file instead of actually writing to the file and reading from it. Can you make sure it's at least unbuffered so we don't run into race conditions?

**Comment by rkooo567** [comment]
@edoakes I can just separate write method from `get_file_discovery_content` and test the latter method. (So we will test the output of this method instead of file contents).

**Comment by rkooo567** [comment]
@edoakes I had to make the file write atomic because otherwise Prometheus fails to read json because of race condition. This doesn't affect prometheus mostly (seems like we can still import metrics properly), but it prints lots of false alarms to users. So I made the file write atomic (basically write a temp file and rename it, which is an atomic operation https://stackoverflow.com/questions/7054844/is-rename-atomic). Please leave more comments if you have concerns! I will push it tomorrow otherwise.

**Pull request merged**